Adding Services to federation clientset

nikhiljindal opened the pull request:
Commits:
1. Regenerate the client without any changes to client-gen
2. Update clientgen to add a parameter to specify generating client only for Services v1 object.
3. Regenerate federation_internalclientset
4. Regenerate federation_release_1_3

Second commit is the most important one. Other 3 commits are auto generated by running client-gen.
I have added a command line argument to client-gen that takes in a list of group/version/resource. If a group version is part of this list, then only the resources in this list are included in the client. For other group versions, the existing check of genclient=true in types.go is used.

Other alternatives considered were:
- Update genclient in types.go to mention the clientset name in which it should be included instead of just saying genclient=true (so Services will say genclient=core,federation while all other v1 resources will say genclient=core). This requires a code change in types.go to change a client set.
- Create another types.go which will only include Services and use that to generate federation clientset. This will lead to duplicate Service definition.

cc @caesarxuchao @lavalamp @jianhuiz @mfanjie @kubernetes/sig-cluster-federation 

<!-- Reviewable:start -->

---

This change is [<img src="http://reviewable.k8s.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/25443)

<!-- Reviewable:end -->

nikhiljindal committed:
Updating client-gen to support a user specified override on the resources to be included in generated client

nikhiljindal commented:
Related PR to ensure that federation_clientset is latest: https://github.com/kubernetes/kubernetes/pull/25447

caesarxuchao commented on cmd/libs/go2idl/client-gen/args/args.go:
```diff
@@ -29,6 +29,10 @@ type Args struct {
 	GroupVersionToInputPath map[unversioned.GroupVersion]string
+
+	// Overrides for which resources should be included in the client.
+	IncludeResourceOverrides map[unversioned.GroupVersion][]string
```
How about `IncludedTypeOverrided`? We are use the word "Type" throughout the client-gen.

caesarxuchao replied:
Since the name is derived from the flag which is user-facing, so I agree `IncludedResourcesOverrides` is a better name, but I think it should be `Included` `Resources` `Overrides`.

nikhiljindal replied:
Done. Renamed to IncludedTypesOverrides

caesarxuchao commented on cmd/libs/go2idl/client-gen/main.go:
```diff
@@ -33,13 +33,14 @@ import (
-	fakeClient    = flag.Bool("fake-clientset", true, "when set, client-gen will generate the fake clientset that can be used in tests")
+	test                     = flag.BoolP("test", "t", false, "set this flag to generate the client code for the testdata")
+	inputVersions            = flag.StringSlice("input", []string{"api/", "extensions/", "batch/"}, "group/versions that client-gen will generate clients for. At most one version per group is allowed. Specified in the format \"group1/version1,group2/version2...\". Default to \"api/,extensions/,batch/\"")
+	includeResourceOverrides = flag.StringSlice("include-resource-overrides", []string{}, "list of group/version/resource for which client should be generated. By default, client is generated for all resources which have genclient=true in types.go. This overrides that. For each groupVersion in this list, only the resources mentioned here will be included. The default check of genclient=true will be  ...
```
Is the user going to specify the `Resource` or just the type's name?

nikhiljindal replied:
Type. Updated to say type in the variable name

caesarxuchao commented on cmd/libs/go2idl/client-gen/main.go:
```diff
@@ -52,6 +53,24 @@ func versionToPath(gvPath string, group string, version string) (path string) {
+		}
+		return "", "", invalidFormatErr
+	case 3:
+		return strings.Join(subs[:length-1], "/"), strings.Join(subs[length-1:], "/"), nil
```
The second returned value could just be `subs[length-1]`

nikhiljindal replied:
fixed

caesarxuchao commented:
Thanks @nikhiljindal. I left some comments in the second commit.

nikhiljindal committed:
Regenerating the federation client with updated client gen

nikhiljindal commented:
Rebased after https://github.com/kubernetes/kubernetes/pull/25447 was merged. Also fixed the code as per comments. 

There are just 2 commits now - one to update client-gen to allow included types overrides and the second commit is to regenerate the client by running hack/update-codegen.sh

PTAL.

nikhiljindal commented:
Not sure what happened here.
Test artifacts are empty. Looks like the test didnt even start

@k8s-bot e2e test this issue: #IGNORE

caesarxuchao commented on cmd/libs/go2idl/client-gen/generators/client_generator.go:
```diff
@@ -166,15 +166,32 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
-				continue
+		for n, t := range p.Types {
+			// filter out types which are not included in user specified overrides.
+			resources, ok := includedTypesOverrides[gv]
```
nit: s/resources/types

nikhiljindal replied:
Fixed

nikhiljindal commented:
Made sure that resource is not mentioned anywhere in my first commit :)

caesarxuchao commented:
Thanks. LGTM.

nikhiljindal commented:
Fixed a typo to fix build. Adding back LGTM

nikhiljindal commented:
@k8s-bot unit test this issue: #IGNORE
